scenarios/waymo_motion/waymo_utility.py

```diff
@@ -0,0 +1,2226 @@
+                scenarios_to_animate.append(scenario_dict[scenario_idx])
+
+            if len(scenarios_to_animate) > 0:
+                plot_scenarios(scenarios_to_animate, True)
+            else:
+                print("No animations were shown as no filter tags matched\n")
+
+        elif re.compile(
+            "^tag([\s]+imported)?[\s]+(all|(?:\s*(\d+))+)$", flags=re.IGNORECASE
+        ).match(user_input):
+            input_lst = user_input.lower().split()
+            imported = True if "imported" in input_lst else False
+            if "all" in input_lst:
+                valid_indexes = [i + 1 for i in range(len(scenario_ids))]
+            else:
+                if imported:
+                    indexes_input = input_lst[2:]
+                else:
+                    indexes_input = input_lst[1:]
+
+                # Check if index passed is valid
+                valid_indexes = check_index_validity(
+                    indexes_input, len(scenario_ids), "tag"
+                )
+            if len(valid_indexes) == 0:
+                continue
+            print(
+                "What Tags do you want to add?\n"
+                "Your response should have tags that are alphanumerical and can have special characters but need to be separated by Comma.\n"
+            )
+            stripped_response, stop_browser = prompt_tags()
+            if stop_browser:
+                return True
+
+            tags = [tag.strip() for tag in stripped_response.lower().split(",")]
+            if imported:
+                for i in range(len(valid_indexes)):
+                    scenario_idx = scenario_ids[valid_indexes[i] - 1]
+                    if scenario_idx in imported_tfrecord_tags:
+                        imported_tfrecord_tags[scenario_idx].extend(
+                            [
+                                tag
+                                for tag in tags
+                                if tag not in imported_tfrecord_tags[scenario_idx]
+                            ]
+                        )
+                    else:
+                        imported_tfrecord_tags[scenario_idx] = tags
+                print("Tags added to `Imported Tags` list")
+
+            else:
+                for i in range(len(valid_indexes)):
+                    scenario_idx = scenario_ids[valid_indexes[i] - 1]
+                    tfrecord_tags[scenario_idx].extend(
+                        [tag for tag in tags if tag not in tfrecord_tags[scenario_idx]]
+                    )
+                print("Tags added to `Tags Added` list")
+
+            time.sleep(1)
+            display_scenarios_in_tfrecord(
+                tfrecord,
+                scenario_dict,
+                tfrecord_tags,
+                imported_tfrecord_tags,
+            )
+            print_commands = True
+
+        elif re.compile(
+            "^untag([\s]+imported)?[\s]+(all|(?:\s*(\d+))+)$", flags=re.IGNORECASE
+        ).match(user_input):
```

Comment:
This is not critical, but I think this code would benefit from a refactor towards a more traditional parser architecture (like in a compiler/interpreter for a programming language). For instance, you could design a "grammar" of allowable commands for each level, and then have a single `parse` function that handles user input in a unified way. This would avoid the heavy use of regex and duplicated string handling code, and allow for pulling common parts into functions.